### PR TITLE
Add the Fides Optimizer

### PR DESCRIPTION
**Pull request by roecla**
The documentation of Fides can be found [here](https://fides-optimizer.readthedocs.io/en/latest/).

### CODE REVIEW

File: src/estimagic/optimization/__init__.py [hunk 1]
```diff
@@ -55,11 +57,15 @@
...
 
+if IS_FIDES_INSTALLED:
+    AVAILABLE_ALGORITHMS["fides"] = fides_optimizers.fides
+
 GLOBAL_ALGORITHMS = [
     "nlopt_direct",
     "nlopt_esch",
     "nlopt_isres",
     "nlopt_crs2_lm",
+    "fides",
```
[the remark this answers is not in the document]
Comment (roecla) [reply src/estimagic/optimization/__init__.py:68]: True. However, it does require bounds which is for which this list is used in the test cases. That's why I put it on the list. 
I would propose to rename the `GLOBAL_ALGORITHMS`, maybe `BOUND_NEEDING_ALGORITHMS`?

File: src/estimagic/optimization/fides_optimizers.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        )
+
+    algo_info = {
+        "primary_criterion_entry": "value",
+        "parallelizes": False,
+        "needs_scaling": False,
+        "name": "fides",
+    }
+
+    def fun(x):
```
[the remark this answers is not in the document]
Comment (roecla) [reply src/estimagic/optimization/fides_optimizers.py:43]: You are right. That is not very efficient. Your proposal is much better.

File: src/estimagic/optimization/fides_optimizers.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import logging
+
+from estimagic.config import IS_FIDES_INSTALLED
+
+if IS_FIDES_INSTALLED:
+    from fides import hessian_approximation
+    from fides import Optimizer
+
+
+def fides(
```
[the remark this answers is not in the document]
Comment (roecla) [reply src/estimagic/optimization/fides_optimizers.py:12]: Somehow, I missed these completely :see_no_evil: but I was able to implement everything supported through the [Options Dictionary](https://fides-optimizer.readthedocs.io/en/latest/_modules/fides/constants.html#Options).

File: src/estimagic/optimization/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (janosg) [reply src/estimagic/optimization/__init__.py:68]: They have an unbounded test case: https://github.com/fides-dev/fides/blob/f481387a3e01a1480c556a826cbdaa9a99636c34/tests/test_minimize.py#L135

Are you sure the algorithm needs finite bounds on all parameters?

File: src/estimagic/optimization/fides_optimizers.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    *,
+    hessian_update_strategy="bfgs",
+    convergence_absolute_criterion_tolerance=CONVERGENCE_ABSOLUTE_CRITERION_TOLERANCE,
+    convergence_relative_criterion_tolerance=CONVERGENCE_RELATIVE_CRITERION_TOLERANCE,
+    convergence_absolute_params_tolerance=CONVERGENCE_ABSOLUTE_PARAMS_TOLERANCE,
+    convergence_absolute_gradient_tolerance=CONVERGENCE_ABSOLUTE_GRADIENT_TOLERANCE,
+    convergence_relative_gradient_tolerance=CONVERGENCE_RELATIVE_GRADIENT_TOLERANCE,
+    stopping_max_iterations=STOPPING_MAX_ITERATIONS,
+    stopping_max_seconds=np.inf,
+    trustregion_initial_radius=1.0,
```
Comment (janosg) [inline src/estimagic/optimization/fides_optimizers.py:33]: Don't we have a default for initial trust region radius?

**Review by janosg — COMMENTED** [review]
Looks good now. 

I am still missing a way to set different trustregion solvers. Is this not possible in fides?

File: src/estimagic/optimization/fides_optimizers.py (hunk 4 above)
Comment (roecla) [reply src/estimagic/optimization/fides_optimizers.py:33]: I did not find it in `algo_options.py` so I decided to hard code the Fides default.

File: src/estimagic/optimization/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (roecla) [reply src/estimagic/optimization/__init__.py]: Apparently pip gave me a rather old version (0.5) which required bounds. Updating to 0.6.3 resolved the problem. I make sure we require Fides to be at least at 0.6.3 now.

**Comment by roecla** [comment]
Tests currently fail because Fides does not find the correct solution in the case, where two parameters have binding bounds. I filed a [Fides issue](https://github.com/fides-dev/fides/issues/43) for this.

**Comment by roecla** [comment]
> I am still missing a way to set different trustregion solvers. Is this not possible in fides?

I implemented everything that is allowed in their [Options](https://fides-optimizer.readthedocs.io/en/latest/generated/fides.constants.html#fides.constants.Options) object. Trying to pass anything that is not in there would raise an Error. 

Their Options has an entry `SUBSPACE_DIM` that is also called the subproblem solver. However, that only specifies the subspace dimensionality (full, 2D or Steihaug's method). Maybe that's what you are looking for?

**Review by janosg — APPROVED** [review]
Very nice, thanks!

**Pull request merged**